File: internal/rules/changedattrs.go
package rules

import (
	"strings"

	"github.com/foadtalsi/tf-predeploy-firewall/internal/parser"
)

// ChangedAttrKey identifies one attribute inside the PR's own .tf diff,
// either top-level ("engine") or inside a nested block
// ("root_block_device.volume_type"). Used by plan-based rules to tell
// whether a value the plan says changed was actually touched by this PR,
// versus having drifted from some other source (console edits, another
// pipeline, a provider default shifting).
type ChangedAttrKey string

// changedAttrsForResource returns the set of attribute keys whose literal
// value differs between base and head, or whose presence changed
// (added/removed). Non-literal expressions (var.x, data.foo.bar) are
// treated conservatively as "changed" — we can't prove they didn't, so we'd
// rather under-report drift than over-report it.
func changedAttrsForResource(head, base *parser.Resource) map[ChangedAttrKey]bool {
	changed := map[ChangedAttrKey]bool{}
	if head == nil || base == nil {
		return changed
	}

	diffAttrMaps(head.Attributes, base.Attributes, "", changed)

	blockByType := func(blocks []*parser.NestedBlock) map[string]*parser.NestedBlock {
		m := map[string]*parser.NestedBlock{}
		for _, b := range blocks {
			m[b.Type] = b
		}
		return m
	}
	headBlocks := blockByType(head.Blocks)
	baseBlocks := blockByType(base.Blocks)
	for blockType, headBlk := range headBlocks {
		baseBlk, ok := baseBlocks[blockType]
		if !ok {
			continue // whole block is new; not a per-attribute drift comparison
		}
		diffAttrMaps(headBlk.Attributes, baseBlk.Attributes, blockType+".", changed)
	}

	return changed
}

// bareResourceAddress strips a terraform plan address down to the bare
// "type.name" form the HCL parser produces (parser.Resource.Address()).
// A plan address may carry a module path prefix
// ("module.vpc.module.subnets.aws_subnet.private") and/or a count/for_each
// instance key suffix ("aws_instance.web[0]", `aws_instance.web["prod"]`),
// neither of which the static HCL scan knows about — it only ever sees the
// resource block's own type and label. Without this normalization, every
// resource inside a module (the overwhelmingly common real-world layout)
// or behind count/for_each would fail to match against changedAttrs and be
// misreported as drift.
func bareResourceAddress(planAddr string) string {
	addr := planAddr
	if idx := strings.IndexByte(addr, '['); idx >= 0 {
		// An instance key suffix only ever appears on the final segment,
		// so it's safe to strip before splitting on ".".
		closeIdx := strings.LastIndexByte(addr, ']')
		if closeIdx > idx {
			addr = addr[:idx] + addr[closeIdx+1:]
		} else {
			addr = addr[:idx]
		}
	}
	parts := strings.Split(addr, ".")
	if len(parts) < 2 {
		return addr
	}
	return parts[len(parts)-2] + "." + parts[len(parts)-1]
}

func diffAttrMaps(head, base map[string]*parser.Attribute, prefix string, out map[ChangedAttrKey]bool) {
	for name, headAttr := range head {
		baseAttr, existed := base[name]
		if !existed {
			out[ChangedAttrKey(prefix+name)] = true
			continue
		}
		if !headAttr.IsLiteral || !baseAttr.IsLiteral || headAttr.RawValue != baseAttr.RawValue {
			out[ChangedAttrKey(prefix+name)] = true
		}
	}
	for name := range base {
		if _, stillPresent := head[name]; !stillPresent {
			out[ChangedAttrKey(prefix+name)] = true
		}
	}
}

File: src/tfpdf/rules/changedattrs.py
"""Identify attributes actually changed by this pull request's Terraform diff."""

from __future__ import annotations

from ..parser import Attribute, NestedBlock, Resource

#: One attribute inside the PR's own diff, either top-level ("engine") or
#: inside a nested block ("root_block_device.volume_type").
ChangedAttrKey = str


def changed_attrs_for_resource(head: Resource | None, base: Resource | None) -> set[ChangedAttrKey]:
    """Return attributes whose literal values or presence differ between base and head."""
    changed: set[ChangedAttrKey] = set()
    if head is None or base is None:
        return changed

    _diff_attr_maps(head.attributes, base.attributes, "", changed)

    head_blocks = _blocks_by_type(head.blocks)
    base_blocks = _blocks_by_type(base.blocks)
    for block_type, head_blk in head_blocks.items():
        base_blk = base_blocks.get(block_type)
        if base_blk is None:
            continue  # whole block is new; not a per-attribute drift comparison
        _diff_attr_maps(head_blk.attributes, base_blk.attributes, block_type + ".", changed)

    return changed


def _blocks_by_type(blocks: list[NestedBlock]) -> dict[str, NestedBlock]:
    return {block.type: block for block in blocks}


def _diff_attr_maps(
    head: dict[str, Attribute],
    base: dict[str, Attribute],
    prefix: str,
    out: set[ChangedAttrKey],
) -> None:
    for name, head_attr in head.items():
        base_attr = base.get(name)
        if base_attr is None:
            out.add(prefix + name)
            continue
        if (
            not head_attr.is_literal
            or not base_attr.is_literal
            or head_attr.raw_value != base_attr.raw_value
        ):
            out.add(prefix + name)
    for name in base:
        if name not in head:
            out.add(prefix + name)


def bare_resource_address(plan_addr: str) -> str:
    """Reduce a plan address to the type.name form used by the HCL parser."""
    addr = plan_addr
    index = addr.find("[")
    if index >= 0:
        # An instance key suffix only ever appears on the final segment, so it
        # is safe to strip before splitting on ".".
        close_idx = addr.rfind("]")
        addr = addr[:index] + addr[close_idx + 1 :] if close_idx > index else addr[:index]
    parts = addr.split(".")
    if len(parts) < 2:
        return addr
    return parts[-2] + "." + parts[-1]
